### PR TITLE
Add SendCavcRemandProcessedLetterTask 

### LINKED ISSUE #15076: Litigation Support: Create new `SendCavcRemandProcessedLetterTask` type
## User or job story
User story: As a Litigation Support, when I have completed the court remand kickoff form and have selected either a JMR or JMPR remand on the form, I need a "SendCavcRemandProcessedLetterTask" task to be created, so that I can take necessary action before sending the letter to an appellant. 

## Acceptance criteria
- [ ] This feature should be accessible to the following user groups: CAVC Litigation Support
- [ ] Consider better naming of SendCavcRemandProcessedLetterTask
- [ ] SendCavcRemandProcessedLetterTask is created on the new appeal stream as a child of the CAVC task
- [ ] SendCavcRemandProcessedLetterTask is assigned to the CAVC Litigation Support organization as unassigned
- [ ] The Actions drop-down has the following available for the Admin of the organization: 
- Assign to person
- [ ] Once assigned by the Admin of the organization, the Actions drop-down has the following available for the assigned user: 
- Add Admin Action
- Mark task complete
  - Indicates letter has been sent to veteran 

### Designs

[➡️ UI Designs and flow ⬅️
](https://www.figma.com/file/8noUqRbvCJzdpwkSNCb8lZ/CAVC-Remand-Designs?node-id=913%3A0)

### PR DESCRIPTION
**Pull request by yoomlam**
Resolves #15076

### Description
Add SendCavcRemandProcessedLetterTask when CavcTask is created.

## Acceptance criteria
- [ ] This feature should be accessible to the following user groups: CAVC Litigation Support
- [ ] Consider better naming of SendCavcRemandProcessedLetterTask 
   - couldn't think of anything better: `SendCavcProcessedLetterTask`, `SendCavcCompleteLetterTask`
- [x] SendCavcRemandProcessedLetterTask is created on the new appeal stream as a child of the CAVC task
- [x] SendCavcRemandProcessedLetterTask is assigned to the CAVC Litigation Support organization as unassigned
- [x] The Actions drop-down has the following available for the Admin of the organization: 
   - Assign to person
- [ ] Once assigned by the Admin of the organization, the Actions drop-down has the following available for the assigned user: 
   - ~Add Admin Action~ Looks like we'll be removing this in #15286, so leaving it out.
   - Mark task complete (Indicates letter has been sent to veteran)

I implemented task reassignment, but only for the user-task (task assigned to a user) not the org-task (task assigned to the CAVC Lit Support org). Reassignment of the org-task will cancel the existing org-task and create a user-task without an org-task parent, which is probably not what we want.

### Testing Plan
1. Create the SendCavcRemandProcessedLetterTask task
```ruby
# Find an appeal with an open DistributionTask to play with
appeal=DistributionTask.open.first.appeal
appeal.treee

distr_task=appeal.tasks.open.where(type: DistributionTask.name).first 
cavc_task=CavcTask.create(appeal: appeal, parent: distr_task)
appeal.reload.treee

send_task=SendCavcRemandProcessedLetterTask.create(appeal: appeal, parent: cavc_task)
appeal.reload.treee
```
1. Sign in as # CAVC_LIT_SUPPORT_ADMIN (VACO) and switch to their [CAVC Lit Support team view](http://localhost:3000/organizations/cavc-lit-support?tab=unassignedTab&page=1). The appeal should be listed in the Unassigned tab.
1. Click on the case; assign task to a CAVC Lit Support team member; check that case is now under the "Assigned" tab.
1. Sign in as the user you assigned it to and go to their queue; click on the case.
1. Play around with task actions, check the case appears in correct queue tab, and check the task tree `appeal.reload.treee`
   * ~ADD_ADMIN_ACTION; cancel it (`Task.where(id: [3374, 3375]).map(&:cancelled!)`) or complete it as the assigned user to have the case assigned back to the assigner. Can also destroy the tasks (`Task.where(id: [3374, 3375]).map(&:destroy)`).~
   * REASSIGN_TO_PERSON
   * MARK_COMPLETE will complete all the way up the tree up to (and including CavcTask since it has no open children tasks).  Can test scenario where CavcTask does have other open children after marking the task complete.
```ruby
                                                      ┌──────────────────────────────────────────────────────────────────────────────────────────────┐
Appeal 9 (evidence_submission) ────────────────────── │ ID   │ STATUS    │ ASGN_BY                │ ASGN_TO                │ UPDATED_AT              │
└── RootTask                                          │ 24   │ on_hold   │                        │ Bva                    │ 2020-11-03 21:30:55 UTC │
    └── DistributionTask                              │ 25   │ on_hold   │                        │ Bva                    │ 2020-11-03 21:30:55 UTC │
        ├── EvidenceSubmissionWindowTask              │ 26   │ assigned  │                        │ MailTeam               │ 2020-11-03 21:30:55 UTC │
        └── CavcTask                                  │ 3369 │ completed │                        │ Bva                    │ 2020-11-04 18:18:39 UTC │
            └── SendCavcRemandProcessedLetterTask     │ 3372 │ completed │                        │ CavcLitigationSupport  │ 2020-11-04 18:18:39 UTC │
                └── SendCavcRemandProcessedLetterTask │ 3373 │ completed │ CAVC_LIT_SUPPORT_ADMIN │ CAVC_LIT_SUPPORT_USER1 │ 2020-11-04 18:18:38 UTC │
                                                      └──────────────────────────────────────────────────────────────────────────────────────────────┘
```

### User Facing Changes
 - [ ] Screenshots of UI changes added to PR & Original Issue

 BEFORE|AFTER
 ---|---
(None) | For CAVC Lit Support Admin user for an org-task image
(None) | For CAVC Lit Support Non-Admin user for a user-task image

### Code Documentation Updates
- [x] Add or update code comments at the top of the class, module, and/or component.

### CODE REVIEW

File: spec/models/tasks/cavc_task_spec.rb [hunk 1]
```diff
@@ -1,42 +1,21 @@
...
-        expect(new_task.errors.messages[:parent]).to be_empty
-
-        expect(appeal.tasks).to include new_task
-        expect(parent_task.children).to include new_task
-
-        expect(new_task.assigned_to).to eq Bva.singleton
-        expect(new_task.label).to eq "All CAVC-related tasks"
-        expect(new_task.default_instructions).to eq [COPY::CAVC_TASK_DEFAULT_INSTRUCTIONS]
-      end
-    end
```
Comment (yoomlam) [inline spec/models/tasks/cavc_task_spec.rb:22]: Replaced by line 12: `it_behaves_like "task requiring specific parent"`

File: spec/models/tasks/cavc_task_spec.rb [hunk 2]
```diff
@@ -1,42 +1,21 @@
...
-
-    context "parent is nil" do
-      let(:parent_task) { nil }
-      it "fails to create task" do
-        new_task = subject
-        expect(new_task.invalid?)
-        expect(new_task.errors.messages[:parent]).to include("can't be blank")
-      end
```
Comment (yoomlam) [inline spec/models/tasks/cavc_task_spec.rb:39]: Replaced by line 12: `it_behaves_like "task requiring specific parent"`

File: app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -208,7 +208,8 @@ def valid_task_classes
     additional_task_classes = Hash[
       *MailTask.subclasses.map { |subclass| [subclass.to_s.to_sym, subclass] }.flatten,
       *HearingAdminActionTask.subclasses.map { |subclass| [subclass.to_s.to_sym, subclass] }.flatten,
-      *ColocatedTask.subclasses.map { |subclass| [subclass.to_s.to_sym, subclass] }.flatten
+      *ColocatedTask.subclasses.map { |subclass| [subclass.to_s.to_sym, subclass] }.flatten,
+      SendCavcRemandProcessedLetterTask.to_s.to_sym, SendCavcRemandProcessedLetterTask
```
Comment (yoomlam) [inline app/controllers/tasks_controller.rb:212]: Needed for task reassignment to work.

File: spec/factories/task.rb [hunk 4]
```diff
@@ -2,9 +2,9 @@
 
 FactoryBot.define do
   module FactoryBotHelper
-    def self.find_first_task_or_create(appeal, task_type)
+    def self.find_first_task_or_create(appeal, task_type, **kwargs)
       (appeal.tasks.open.where(type: task_type.name).first if appeal) ||
-        FactoryBot.create(task_type.name.underscore.to_sym, appeal: appeal)
+        FactoryBot.create(task_type.name.underscore.to_sym, appeal: appeal, **kwargs) { |t| yield(t) if block_given? }
```
Comment (yoomlam) [inline spec/factories/task.rb:7]: `{ |t| yield(t) if block_given? }` is not used in this PR but could be useful.

File: app/repositories/task_action_repository.rb [hunk 5]
```diff
@@ -75,12 +75,13 @@ def assign_to_hearings_user_data(task, user = nil)
...
     def assign_to_user_data(task, user = nil)
       users = potential_task_assignees(task)
-
       extras = if task.is_a?(HearingAdminActionTask)
                  {
                    redirect_after: "/organizations/#{HearingsManagement.singleton.url}",
                    message_detail: COPY::HEARING_ASSIGN_TASK_SUCCESS_MESSAGE_DETAIL
                  }
+               elsif task.is_a?(SendCavcRemandProcessedLetterTask) && task.assigned_to_type == "Organization"
+                 { redirect_after: "/organizations/#{CavcLitigationSupport.singleton.url}" }
```
Comment (yoomlam) [inline app/repositories/task_action_repository.rb:84]: Stay in the "team cases" view after assigning a case.

File: app/models/tasks/cavc_task.rb [hunk 6]
```diff
@@ -12,7 +12,7 @@ class CavcTask < Task
   before_validation :set_assignee
 
   def self.label
-    "All CAVC-related tasks"
+    COPY::CAVC_TASK_LABEL
```
Comment (hschallhorn) [inline app/models/tasks/cavc_task.rb:15]: 👌

File: app/models/tasks/send_cavc_remand_processed_letter_task.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+##
+# Task for Litigation Support to take necessary action before sending the CAVC-remand-processed letter to an appellant.
+# This task is for CAVC Remand appeal streams.
+# Expected parent: CavcTask
+# Expected assigned_to.type: User
+
+class SendCavcRemandProcessedLetterTask < Task
+  validates :parent, presence: true,
+                     parentTask: { task_types: [CavcTask, SendCavcRemandProcessedLetterTask] },
```
Comment (hschallhorn) [inline app/models/tasks/send_cavc_remand_processed_letter_task.rb:11]: 👏

File: app/repositories/task_action_repository.rb (hunk 5 above)
Comment (hschallhorn) [reply app/repositories/task_action_repository.rb:84]: I wonder if a better play here would be to send the user back to the organization queue for all tasks that are assigned to an org
```ruby
extras[:redirect_after] = "/organizations/#{task.assigned_to.url}" if task.assigned_to_type == Organization.name
```
Would cover both redirects for these special cases, but also provide redirecting that makes more sense when assigning any org task.
Thoughts? Out of scope of this pr?

File: spec/factories/task.rb [hunk 8]
```diff
@@ -359,6 +359,11 @@ def self.find_first_task_or_create(appeal, task_type)
         parent { FactoryBotHelper.find_first_task_or_create(appeal, DistributionTask) }
       end
 
+      factory :send_cavc_remand_processed_letter_task, class: SendCavcRemandProcessedLetterTask do
+        assigned_to { CavcLitigationSupport.singleton }
+        parent { FactoryBotHelper.find_first_task_or_create(appeal, CavcTask) }
```
Comment (hschallhorn) [inline spec/factories/task.rb:364]: 👌

File: spec/feature/queue/cavc_task_queue_spec.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+RSpec.feature "CAVC-related tasks queue", :all_dbs do
+  let!(:org_admin) do
+    create(:user, full_name: "Adminy CacvRemandy") do |u|
+      CavcLitigationSupport.singleton.add_user(u)
+      OrganizationsUser.make_user_admin(u, CavcLitigationSupport.singleton)
```
Comment (hschallhorn) [inline spec/feature/queue/cavc_task_queue_spec.rb:7]: `make_user_admin` handles the addition of the user to the org as well!
https://github.com/department-of-veterans-affairs/caseflow/blob/43562ed9e75b14c6f949802521da9df6a4214c2b/app/models/organizations_user.rb#L16-L21

File: spec/models/tasks/send_cavc_remand_processed_letter_task_spec.rb [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+      end
+    end
+  end
+
+  describe "FactoryBot.create(:send_cavc_remand_processed_letter_task) with different arguments" do
+    context "appeal is provided" do
+      let(:appeal) { create(:appeal) }
+      let!(:cavc_task) { create(:cavc_task, appeal: appeal) }
+      let!(:send_task) { create(:send_cavc_remand_processed_letter_task, appeal: appeal) }
+      it "finds existing parent_task to use as parent" do
```
Comment (hschallhorn) [inline spec/models/tasks/send_cavc_remand_processed_letter_task_spec.rb:42]: Seems like we should be testing the parent of the send task in these 3 tests, no?

File: spec/models/tasks/send_cavc_remand_processed_letter_task_spec.rb [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+      end
+    end
+  end
+
+  SendCRPLetterTask = SendCavcRemandProcessedLetterTask
+  describe "#available_actions" do
+    let(:org_admin) do
+      create(:user) do |u|
+        CavcLitigationSupport.singleton.add_user(u)
+        OrganizationsUser.make_user_admin(u, CavcLitigationSupport.singleton)
```
Comment (hschallhorn) [inline spec/models/tasks/send_cavc_remand_processed_letter_task_spec.rb:78]: Same as the [above](https://github.com/department-of-veterans-affairs/caseflow/pull/15548/files?file-filters%5B%5D=.rb#r519935387)!

**Review by hschallhorn — COMMENTED** [review]
Ran testing from a `make reset`. Kat added some seeds and, with your additions, we've got 9 open `SendCavcRemandProcessedLetterTask`s to play with!

Screen Shot 2020-11-09 at 12 02 40 PM

File: app/models/tasks/send_cavc_remand_processed_letter_task.rb [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+                     on: :create
+
+  before_validation :set_assignee
+
+  USER_ACTIONS = [
+    Constants.TASK_ACTIONS.ADD_ADMIN_ACTION.to_h,
```
Comment (hschallhorn) [inline app/models/tasks/send_cavc_remand_processed_letter_task.rb:17]: Looks like we'll be removing this in #15286. Should we add this here at all?

**Review by hschallhorn — COMMENTED** [review]
Checked bulk task assign as well! Looks great :trulyamazingpanda:

File: app/repositories/task_action_repository.rb (hunk 5 above)
Comment (yoomlam) [reply app/repositories/task_action_repository.rb]: Using your suggestion, [2 tests failed](https://app.circleci.com/pipelines/github/department-of-veterans-affairs/caseflow/36462/workflows/8f3fbe87-dc78-4d3d-9bdc-9977851e3c01/jobs/135743). I think a separate ticket is appropriate.

File: app/models/tasks/send_cavc_remand_processed_letter_task.rb (hunk 12 above)
Comment (yoomlam) [reply app/models/tasks/send_cavc_remand_processed_letter_task.rb:17]: Removing.

File: spec/models/tasks/cavc_task_spec.rb [hunk 13]
```diff
@@ -50,6 +29,7 @@
         expect(RootTask.count).to eq 1
         expect(DistributionTask.count).to eq 1
         expect(CavcTask.count).to eq 1
+        expect(cavc_task.parent).to eq parent_task
```
Comment (yoomlam) [inline spec/models/tasks/cavc_task_spec.rb:32]: Check for expected parent task in these 3 tests.

**Review by hschallhorn — APPROVED** [review]
Looks great!

**Pull request merged**